sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/channel/DefaultChannel.java

```diff
@@ -384,22 +385,50 @@ private synchronized void triggerIngestion(final String batchId, final GroupStat
             /* Remember this batch. */
             groupState.mSendingBatches.put(batchId, batch);
 
+            /*
+             * Due to bug on old Android versions (verified on 4.0.4),
+             * if we start an async task from here, i.e. the async persistence handler thread,
+             * we end up with AsyncTask configured with the wrong Handler to use for onPostExecute
+             * instead of using main thread as advertised in Javadoc (and its a static field there).
+             *
+             * Our SDK guards against an application that would make a first async task in non UI
+             * thread before SDK is initialized, but we should also avoid corrupting AsyncTask
+             * with our wrong handler to avoid creating bugs in the application code since we are
+             * a library.
+             *
+             * So make sure we execute the async task from UI thread to avoid any issue.
+             */
+            HandlerUtils.runOnUiThread(new Runnable() {
+
+                @Override
+                public void run() {
+                    sendLogs(groupState, stateSnapshot, batch, batchId);
+                }
+            });
+        }
+    }
+
+    /**
+     * Send logs.
+     */
+    private synchronized void sendLogs(final GroupState groupState, final int stateSnapshot, List<Log> batch, final String batchId) {
```

Comment:
Every single we use synchronized, it's at method level to avoid code coverage issue complaining we don't test exiting a synchronized block with an exception.